src/harness/reference_models/fss_purge/fss_purge.py

```diff
@@ -0,0 +1,318 @@
+                                  weight_2=fss_weight_2)
+  return fss_entity
+
+def getNeighboringCbsdsToFss(cbsd_list, fss_record):
+  """ Get the list of all CBSDs that lie within 40 KMs of FSS.
+
+  Args:
+    cbsd_list:  List of CbsdData dictionaries as defined in the SAS-SAS specification.
+    fss_record: A FSS record dictionary.
+  Returns:
+    List of CBSDs lying within 40 KMs of FSS.
+  """
+
+  neighboring_cbsds = []
+  for cbsd in cbsd_list:
+    distance_km, _, _ = vincenty.GeodesicDistanceBearing(
+        fss_record['deploymentParam'][0]['installationParam']['latitude'],
+        fss_record['deploymentParam'][0]['installationParam']['longitude'],
+        cbsd['registrationRequest']['installationParam']['latitude'],
+        cbsd['registrationRequest']['installationParam']['longitude'])
+
+    if distance_km <= 40:
+        neighboring_cbsds.append(cbsd)
+  return neighboring_cbsds
+
+def performPurge(cbsds, fss_entity):
+  """ Implementation of algorithm described in R2-SGN-29.
+
+  The list of grants to be purged based on the interference caused by their
+  out of band emission are identified and the grants are removed from the CBSDs.
+
+  Args:
+    cbsds: List of CbsdData objects that are in the neighborhood to an FSS.
+    fss_entity: The FssProtectionPoint tuple of the FSS for which the purge is to be performed.
+  """
+
+  # Calculate the OOBE interference value for each CBSD based on the grant with the highest frequency.
+  calculateOobeInterference(cbsds, fss_entity)
+
+  # Sort the CBSDs based on the calculated interference value.
+  sorted_cbsds = sorted(cbsds, key=lambda cbsd: cbsd['oobe_interference'])
+  # Calculate the threshold value in dBm for the OOBE interference
+  oobe_threshold_value = -129 + 10 * (np.log10(REF_BW/interference.ONE_MHZ) - FSS_OOBE_MARGIN)
+
+  # Find the largest index such that aggregation of the interferences do not cross 
+  # the OOBE threshold
+  oobe_interferences = [interference.dbToLinear(cbsd['oobe_interference']) for cbsd in sorted_cbsds]
+  cumulated_interference = np.cumsum(oobe_interferences)
+  index = np.searchsorted(cumulated_interference, interference.dbToLinear(oobe_threshold_value))
+  cbsds_to_purge = sorted_cbsds[index:]
+  sorted_cbsds = sorted_cbsds[:index]
+
+  # Remove the grants from the CBSDs that are in the purge list.
+  if cbsds_to_purge:
+    for index, cbsd in enumerate(cbsds_to_purge):
+      if cbsd['grants']:
+        # Get the grant that was used for the OOBE interference calculation, that
+        # is the grant with highest highFrequency.
+        grant_with_max_high_freq = max(cbsd['grants'],\
+              key=lambda x: x['operationParam']['operationFrequencyRange']['highFrequency'])
+
+        # Get all the grants that share the same emission mask as the one that 
+        # was used for OOBE calculation
+        if (grant_with_max_high_freq['operationParam']['operationFrequencyRange']
+            ['highFrequency'] > 3690000000):

```

Comment:
The logic here assumes that the FSS band type for the TT&C type starts at 3700. This will be controlled by the input configuration.Do we need to have some code that would validate the input and ensure that the inputs conform to this assumption? Otherwise, the logic here would not work.

Reply:
As per update from WinnF WG.4,   purge list will be only calculated for 3700MHz -3705 MHz (for the FSS earth   station with TT&C designation).